### PR TITLE
Improved the way the Settings Registry can handle stacks/arrays.

### PR DESCRIPTION
**Pull request by AMZN-koppersr**
The ConfigurableStack makes configuring stacks and arrays through the Settings Registry easier than using direct serialization to a container like AZStd::vector. It does this by using JSON Objects rather than JSON arrays, although arrays are supported for backwards compatibility. Two key words were added:
- $stack_before : Insert the new entry before the referenced entry. Referencing is done by name.
- $stack_after  : Insert the new entry after the referenced entry. Referencing is done by name.

to allow inserting new entries at specific locations. An example of a .setreg file at updates existing settings would be:

```
// Original settings
{
    "Settings in a stack":
    {
        "AnOriginalEntry":
        {
            "MyValue": "hello",
            "ExampleValue": 84
        },
        "TheSecondEntry":
        {
            "MyValue": "world"
        }
    }
}

// Customized settings.
{
    "Settings in a stack":
    {
        // Add a new entry before "AnOriginalEntry" in the original document.
        "NewEntry":
        {
            "$stack_before": "AnOriginalEntry",
            "MyValue": 42
        },
        // Add a second entry after "AnOriginalEntry" in the original document.
        "SecondNewEntry":
        {
            "$stack_after": "AnOriginalEntry",
            "MyValue": "FortyTwo".
        },
        // Update a value in "AnOriginalEntry".
        "AnOriginalEntry":
        {
            "ExampleValue": 42
        },
        // Delete the "TheSecondEntry" from the settings.
        "TheSecondEntry" : null,
    }
}
```

The ConfigurableStack uses an AZStd::shared_ptr to store the values. This supports settings up a base class and specifying derived classes in the settings, but requires that the base and derived classes all have a memory allocator associated with them (i.e. by using the "AZ_CLASS_ALLOCATOR" macro) and that the relation of the classes is reflected. Loading a ConfigurableStack can be done using the GetObject call on the SettingsRegistryInterface.

Signed-off-by: AMZN-koppersr <82230785+AMZN-koppersr@users.noreply.github.com>

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Settings/ConfigurableStack.h [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        ~ConfigurableStack() override = default;
+
+        const TypeId& GetNodeType() const override;
+
+        Iterator begin();
+        Iterator end();
+        ConstIterator begin() const;
+        ConstIterator end() const;
+        ConstIterator cbegin() const;
+        ConstIterator cend() const;
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Settings/ConfigurableStack.h:114]: Actually with the new range support. The need to implement cbegin and cend aren't needed.

[std::ranges::cbegin](https://en.cppreference.com/w/cpp/ranges/cbegin) and [std::ranges::cend](https://en.cppreference.com/w/cpp/ranges/cend) will use the `begin` and `end` const member functions if available

File: Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        static void Reflect(AZ::ReflectContext* context)
+        {
+            if (auto sc = azrtti_cast<AZ::SerializeContext*>(context))
+            {
+                sc->Class<ConfigInt>()->Field("Value", &ConfigInt::m_value);
+            }
+        }
+    };
+
+    struct ConfigurableStackTests : public AllocatorsFixture
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp:35]: FYI the `ScopedAllocatorFixture` is the preferred way to initialize the AZ Allocators.

It doesn't require override the Setup/TearDown function to allocate memory after the base class `AllocatorsFixture::SetUp` has been called and to deallocate memory before the `AllocatorsFixture::TearDown` has been called.

It also allows for value types which uses the AZ System Allocator to be members of the fixture.

File: Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            m_jsonRegistrationContext->EnableRemoveReflection();
+            Reflect(m_jsonRegistrationContext);
+            m_jsonRegistrationContext->DisableRemoveReflection();
+
+            m_serializeContext->EnableRemoveReflection();
+            Reflect(m_serializeContext);
+            m_serializeContext->DisableRemoveReflection();
+
+            delete m_jsonRegistrationContext;
+            delete m_serializeContext;
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp:76]: These can be put into a unique_ptr if  this fixture inherits from `ScopedAllocatorFixture`

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp (hunk 2 above)
Comment (hultonha) [reply Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp:35]: I wasn't aware of this either, I do know using `AllocatorsTestFixture` is preferred (I think the old name was just left for compatibility with old code) but `ScopedAllocatorFixture` sounds even better. Maybe this should be added to the testing docs if it's not there already?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp:35]: There are testing docs?
If so they should be added there.
Also can you link me those documents?

Comment (hultonha) [reply Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp:35]: There are! Though I don't think the merge has happened yet. I'd keep an eye on https://github.com/o3de/o3de.org/pull/1394 and maybe see if the recommendation could fit somewhere in there potentially? Might be worth reaching out to @Kadino about it next week.

Comment (Kadino) [reply Code/Framework/AzCore/Tests/Settings/ConfigurableStackTests.cpp:35]: There's already a place for AZTest/GoogleTest docs, which would be the appropriate place to add allocator info: https://www.o3de.org/docs/user-guide/testing/aztest/aztest/

The linked PR is for a getting started guide, which isn't the right place for this detail.

File: Code/Framework/AzCore/AzCore/Settings/ConfigurableStack.h (hunk 1 above)
Comment (AMZN-koppersr) [reply Code/Framework/AzCore/AzCore/Settings/ConfigurableStack.h:114]: Will remove them then as less code to maintain is better.

**Pull request merged**